### ISSUE TITLE
Properties of object defined as `Record<string, never>` can be successfully assigned to any type

### ISSUE THREAD

**Issue by kael-shipman** [microsoft/TypeScript#60265]
### 🔎 Search Terms

never successfully assigned should be error

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about `never`

### ⏯ Playground Link

https://www.typescriptlang.org/play/?ts=5.6.3&ssl=19&ssc=25&pln=1&pc=1#code/C4TwDgpgBAyg9gWwgFXNAvFA3gWAFBSFRwB2EAXFCQK4IBGEATgNz5FTADuclAzsIwCWJAOasCRYAAtGEClDpw4AGwgBDEuPYAzONUYB+SgCUIAYziMAJgB5+Q0QBoo1EgGsScTiQB84gL74+BYk-ApqVibmlrb2wiLOZABuTD5QmFiBePgA9DlQyFKCvFC8UnrKVhwyXlBMjJa8zrzCZtCeZMTa1RC80GANkIzAgr11AB7FwFAaVdJMELqyUGYantPa1MragsrKPVDCwEzaam3BpGG8iBDS8ZTwSKiQ6dhsRKTydBEAdJ+O70IXB44SsP2BAIkQJkcko3zB0lkEEhOj0jDhv10+khWVy+QAEl4IClGM55stOGoSp4OGheGYhGBpvVLDMSFUAJIAcgQVFq2iRHDgLj61WKHEESEU9Hw1yQd1E4O4P14YGUgmAAAoAEQ5bUASiAA

### 💻 Code

```ts
type SomeType = {
    one: number;
    two: string;
    three: boolean;
    four?: Record<string, unknown>;
}

const bad: Record<string, never> = {}

// This should throw errors, since none of these properties exist and therefore cannot fulfill the interface
const something: SomeType = {
    one: bad.one,
    two: bad.two,
    three: bad.three,
    four: bad.four,
}

// However, there was no typescript error and I'm now free to use this timebomb
something.two.split("/")
```

### 🙁 Actual behavior

Typescript happily accepted the input even though it was invalid and this tiny test app blew up trying to split `undefined`

### 🙂 Expected behavior

Should have issued a type error

**Comment by nmain** [comment]
`never` isn't relevant here; the problem is the index access:

https://www.typescriptlang.org/play/?ts=5.6.3#code/C4TwDgpgBAyg9gWwgFXNAvFA3gWAFBSFTADucAXFAM7ABOAlgHYDmA3PgL774DGcjNKACMAhgBNKAJQh9aYgDw0GLADTU6TZgD4omLFzz4A9EajIAFvSrVzcAK4AbMcXO04JKBFpvaVNVSYeaEZ+aDgAMxcIKmgwN0haYHpozwAPK2AoEUZnYHMvCHC4WmgebJDM8Mdw+gcHKKgmYC9wkSDefkEqRAg8zUp4JFRIXWx8ImIySlExADpSOBVObjwTKAAJdwgANy81PIKoEhFrEOI0Kh4GMEyvHyycqABJAHIEKBCPcJLoYDgoOwxFxWYj0JBCRBCfDdJB9FjzMizKhgBz0YAACgAREZMQBKfBAA

The nouncheckedindexedaccess option catches these.

**Comment by kael-shipman** [comment]
@nmain that doesn't sound right to me. The definition of `never` is that it should not be assignable to anything - `never` is not and can never be a `string`, for example. The fact that I can assign a value typed as `never` to a value typed as `string` seems like it would clearly be an error. And yes, enabling that option will make the example successfully fail, but that seems like it's only a technicality of the way I wrote the example, since all it does is add an implicit `| undefined` to all values accessed via an index (although to be fair, I can't think of a good alternative example off the top of my head).

**Comment by dansaper** [comment]
See https://www.typescriptlang.org/play/?ts=5.6.3#code/CYUwxgNghgTiAEYD2A7AzgF3gTwFzxRADcQYBuAWAChl0sAPfTGASxQHN4BeHSm1TPABe+QiRjd49MkA:
```
declare const y: never;
const x: string = y;
const z: never = x; // Type 'string' is not assignable to type 'never'.
```
That is, a variable of type `never` cannot be assigned any value (excepted those of type `never`). However, a variable of any type can always be assigned to a value of type `never`.

**Comment by jcalz** [comment]
"The definition of `never` is that it should not be assignable to anything" No, [the `never` type is assignable to *everything*.](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#exhaustiveness-checking).  It's TypeScript's [bottom type](https://en.wikipedia.org/wiki/Bottom_type):

```ts
declare let nvr: never;
const str: string = nvr; // okay
const num: number = nvr; // okay
const dat: Date = nvr; // okay
```

You've got it exactly backwards; the definition of `never` is that nothing (other than `never`) is assignable to it:

```ts
nvr = str; // error;
nvr = num; // error
nvt = dat; // error
```

As such, TypeScript is behaving as intended here. If you can actually find a value of type `never` (and you shouldn't be able to) then you can assign it to anything else.

**Comment by kael-shipman** [comment]
I understand that that's _true;_ I wonder if you might at least agree that it's alarming that in my example I can easily get to a fatal error with no warning. Or put another way, it seems fairly obvious that the situation I've outlined _should_ produce an error, regardless of the theory behind the type.

**Comment by jcalz** [comment]
Yes, you apparently want `--noUncheckedIndexedAccess`, which was already pointed out. The situation has little to do with `never`.

**Comment by kael-shipman** [comment]
Ok, I'll give up on this, although I _strongly_ disagree that `--noUncheckedIndexAccess` is the solution here. You yourself pointed out that `declare let nvr: never; const str: string = nvr;` works and will lead to the exact same error I produced in my example, and `--noUncheckedIndexAccess` will not save you.

Anyway, I'll let it go for now, as I appreciate that you guys have bigger fish to fry 😁 .

**Comment by jcalz** [comment]
None of us guys frying fish here are official TS team folks so if you disagree both strongly and italically you might wait for them to chime in before giving up.  

I'll try one more time to spell out what's happening; feel free to ignore it.

You can never get your hands on a value of type `never`, so any manipulation you do with a value of that type is indicative that something has *already* gone wrong:

```ts
function foo(): never { throw new Error("oopsie"); }
let nvr: never = foo(); // <-- this is where the runtime error happens
const str: string = nvr; // <-- this code never runs
str.toUpperCase(); // <-- this code never runs 
```

The problem isn't with the assignment from `nvr` to `str`, the problem is that code flow has aborted before you even get there.  That's what `never` is supposed to be, an indication that something impossible has happened and that code flow cannot reach the code you're looking at.

That means value of type `Record<string, never>` is therefore something where reading a property either gives you `undefined` or it throws:

```ts
const rec: Record<string, never> = {
  get x(): never { throw new Error("no x") },
  get y(): never { throw new Error("no y either") },
}
```

So here's where `--noUncheckedIndexedAccess` matters. With it disabled as usual, TypeScript just ignores the possibility that you might get `undefined`, so it imagines that all properties you read are of type `never`, whereas they are very likely to be `undefined`. So TypeScript thinks control flow can't reach your line and will allow you to assign `undefined` to `string` or `number`. The problem is assigning `undefined`, not `never`.  

```ts
// --noUncheckedIndexedAccess disabled
const num: number = rec.z; // okay, but this is a problem because in fact num is undefined.
const undef: undefined = rec.x; // okay, but this doesn't hurt anything because control flow never reaches here
```


If you turn *on* `--noUncheckedIndexedAccess`, then *every* property you read will be treated as type `undefined`, since, if the control flow can actually make it past the read, the property *will* be `undefined`:

```ts
// --noUncheckedIndexedAccess enabled
const num: number = rec.z; // error as expected, Type 'undefined' is not assignable to type 'number'.
const undef: undefined = rec.x; // okay, but this doesn't hurt anything because control flow never reaches here
```

**Comment by kael-shipman** [comment]
That's a good point about not _realistically_ being able to reach a scenario in which this is a problem. I can see how in a fully typescript program that would be the case, and of course in the opposite case of accepting input from the outside world you're already introducing more uncertainty than the simple possibility that we'll run into a situation we didn't think could exist (i.e., `never`). So yes, I can accept that.

Still, I wonder what would happen if they just made `never` not assignable to anything..... Would it blow everything up? Seems like it wouldn't, since it's never supposed to happen anyway 🤷 . And doing so would make it more "right" in my opinion, solving the problem without using an option that (in our case) is not feasible to enable. (As a sidebar, I would love to enable that option but our codebase is already littered with instances of using values directly, and it would be a huge pain to go through and handle them or ignore them with `!`.)

**Comment by guillaumebrunerie** [comment]
You might be confusing `never` with `unknown`? If you want a type that cannot be assigned to anything, it is `unknown` you are looking for.

In terms of sets, `unknown` represents the set of all possible values, while `never` represents the empty set. So it is entirely correct to say that any value of type `never` is also of type `string`, as the empty set is a subset of the set of all strings (or of any other set).

Making `never` behave like `unknown` would completely defeat its purpose, as it would basically mean that only functions returning `unknown` would be allowed to throw exceptions.

**Comment by RyanCavanaugh** [comment]
> Still, I wonder what would happen if they just made never not assignable to anything..... Would it blow everything up? Seems like it wouldn't, since it's never supposed to happen anyway 🤷 . And doing so would make it more "right" in my opinion, solving the problem without using an option that (in our case) is not feasible to enable

There's a ton of misapprehension built into this statement; it's like asking what would happen if you said infinity was less than every other number. Well, it wouldn't be infinity anymore, and then you'd need a new name to give that limit, and then you'd be right back where you started.

The definition of `never` is the empty set of values, and the definition of "assignable to" is "is a subset of values", and the empty set subsets all other sets. I don't know which of those facts you're proposing be thrown out when you say "make `never` not assignable to anything".

**Comment by kael-shipman** [comment]
Again, I'd like to emphasize that I understand and agree that my goal here deviates from the theory. My question is, "should a value that by definition cannot exist be freely assignable to a value that by definition does exist?" Yes, "assignable to" has special meaning in the theory underpinning typescript. But in the same way that the rules were bent to allow index access not reflect that operation's inherent uncertainty (via the `noUncheckedIndexedAccess` setting, which obliterates a hard theoretical fact about index access), I'm proposing that the rules be bent to make this particular aspect of typescript make more sense (and be quantifiably safer, which ultimately is the goal of the language). My understanding is that _precisely because_ this situation is never supposed to happen in the first place, there would be no real consequence to end users other than suddenly discovering upon upgrading their typescript version that something they thought was safe was in fact definitely broken.

**Comment by RyanCavanaugh** [comment]
If you just pencil this out a few steps, you run into huge problems.

The type `string & number` reduces to `never`, because it has zero inhabitants.

The type `string | never` *today* is the same as `never`, because `never` is a subtype of string. But that's no longer true.

So let's say you have a function that filters arrays to just their strings, and you filter it:
```ts
declare function stringsOnly<T>(arr: T[]): (T & string)[];

const arr = [1, "hello"];
const s_arr = stringsOnly(arr);
```
Now the type of `s_arr` isn't `string[]`, it's `(string | never)[]`. Since `never` isn't assignable to `string` anymore, you can't use this as a `string[]`, even though that's what `(T & string)[]` was supposed to do. It's not really clear how you get out of this situation.

**Comment by guillaumebrunerie** [comment]
Why are you so convinced it has something to do with `never`? If anything, it is `Record` that you should blame:
```typescript
const bad: Record<string, string> = {};
// Uh-oh
bad.foo.toLowerCase()
```

Also, where did you get a value of type `Record<string, never>` to begin with? It's usually already pretty rare to have to deal with `never`, so a record of `never` values seems very unusual to me.

**Comment by fatcerberus** [comment]
@guillaumebrunerie `noUncheckedIndexedAccess` prevents the “uh oh” in that example. But `never` is assignable to everything on purpose because it represents the result of an operation that threw an exception.

**Comment by kael-shipman** [comment]
Ok, makes sense, and this answers my question. The operation I'm hoping for would indeed cause broad problems in the implementation. I'm satisfied with that.

To @guillaumebrunerie's comment:

Yes, your example illustrates the "bending of the rules" that I was referencing with the `noUncheckedIndexedAccess: false` setting. Your example _should_ fail, but you can arbitrarily configure typescript to make it not fail.

To your second point about where I got a type of `Record<string, never>`, maybe this is my own misunderstanding, but I have used that type to indicate an object that cannot have any properties. In the case that spawned all this, I was typing a generic request type such that if you didn't explicitly give it a body type, the body type resulted in `Record<string, never>`, which I intended to mean "an empty object". I admit, I grabbed that type out of habit, and you're right that `unknown` would probably have been a more appropriate type in that case and would have solved my problem. I just ended up shocked that there were no errors for something that was so obviously incorrect.

**Comment by fatcerberus** [comment]
TypeScript is actually not type-theoretically airtight by design. It’s designed to strike a balance between soundness and usability in order to model idiomatic JS as closely as possible, so you can easily find holes in it.

**Comment by kael-shipman** [comment]
Totally agree. The setting is always enabled in projects I start from scratch, but for this particular project it's not feasible.

Anyway, I'm satisfied that A) typescript is working according to its theory; and B) making an exception like this would cause widespread problems. Closing this out.

**Comment by MartinJohns** [comment]
> I have used that type to indicate an object that cannot have any properties

You can't represent such a type in TypeScript.
